### PR TITLE
refactor(api): deprecate `ModuleContext.geometry`

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR officially deprecates the internal `ModuleContext.geometry` property for the upcoming PAPI version 2.14 for RCORE-423.

- Internal uses of `ModuleContext.geometry` have been replaced by ProtocolEngine state selectors
- External uses of `ModuleContext.geometry` will be replaced by public ModuleContext methods via RCORE-424

## Changelog

- Deprecate `ModuleContext.geometry` when running on the ProtocolEngine core
- Un-skip various old Protocol API tests

## Review requests

Smoke test a protocol with modules on hardware

## Risk assessment

Low

[RCORE-423]: https://opentrons.atlassian.net/browse/RCORE-423?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RCORE-424]: https://opentrons.atlassian.net/browse/RCORE-424?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #11938** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11938" target="_blank">Graphite</a>  👈
    * **PR #11939** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11939" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11938?utm_source=stack-comment).

**Review by sfoster1 — APPROVED** [review]
This looks good to me. Does gathering the module cleanup stuff help?

File: api/src/opentrons/protocol_api/module_contexts.py [hunk 1]
```diff
@@ -183,16 +189,30 @@ def load_labware_by_name(
     @requires_version(2, 0)
     def labware(self) -> Optional[Labware]:
         """The labware (if any) present on this module."""
-        return self._core.geometry.labware
+        if isinstance(self._core, LegacyModuleCore):
+            return self._core.geometry.labware
```
Comment (mcous) [inline api/src/opentrons/protocol_api/module_contexts.py]: Replace this by implementing the `get_module_item` method instead

File: api/src/opentrons/protocol_api/core/protocol_api/protocol_context.py [hunk 2]
```diff
@@ -403,6 +403,12 @@ def get_slot_item(
         """Get the contents of a given slot, if any."""
         raise NotImplementedError("LegacyProtocolCore.get_slot_item not implemented")
 
+    def get_module_item(
+        self, module_core: legacy_module_core.LegacyModuleCore
+    ) -> Optional[LabwareImplementation]:
+        """Get the item on top of a given module, if any."""
+        raise NotImplementedError("LegacyProtocolCore.get_module_item not implemented")
```
Comment (mcous) [inline api/src/opentrons/protocol_api/core/protocol_api/protocol_context.py]:
```suggestion
    def get_module_item(
        self, module_core: legacy_module_core.LegacyModuleCore
    ) -> Optional[LabwareImplementation]:
        """Get the item on top of a given module, if any."""
        labware = module_core.geometry.labware
        return labware._implementation if labware is not None else None
```

File: api/tests/opentrons/conftest.py [hunk 3]
```diff
@@ -167,8 +167,10 @@ async def _build_ot2_hw() -> AsyncGenerator[ThreadManagedHardware, None]:
         yield hw_sim
     finally:
         config.robot_configs.clear()
-        for m in hw_sim.attached_modules:
-            await m.cleanup()
+        await asyncio.gather(
+            (m.cleanup() for m in hw_sim.wrapped().attached_modules),
+            return_exceptions=True,
+        )
```
Comment (mcous) [inline api/tests/opentrons/conftest.py]: Revert as unnecessary; perf gains are very unlikely vs readability

File: api/tests/opentrons/conftest.py [hunk 4]
```diff
@@ -190,8 +192,10 @@ async def _build_ot3_hw() -> AsyncGenerator[ThreadManagedHardware, None]:
...
```
Comment (mcous) [inline api/tests/opentrons/conftest.py]: Revert

File: api/tests/opentrons/protocol_api/core/engine/test_protocol_core.py [hunk 5]
```diff
@@ -448,6 +456,9 @@ def test_load_module(
     ).then_return(
         LoadedModule.construct(id="abc123")  # type: ignore[call-arg]
     )
+    decoy.when(mock_engine_client.state.labware.get_id_by_module("abc123")).then_raise(
+        LabwareNotLoadedOnModuleError("oh no")
+    )
 
     assert subject.get_slot_item(DeckSlotName.SLOT_1) is result
```
Comment (mcous) [inline api/tests/opentrons/protocol_api/core/engine/test_protocol_core.py:465]: Add missing assert that `subject.get_module_item(module_core) is None`

**Review by mcous — COMMENTED** [review]
A few changes from code review

File: api/src/opentrons/protocol_api/core/engine/protocol.py [hunk 6]
```diff
@@ -359,6 +360,16 @@ def get_slot_item(
 
         return None
 
+    def get_module_item(self, module_core: ModuleCore) -> Optional[LabwareCore]:
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/core/engine/protocol.py]: should this be called `get_labware_on_module`?

**Review by sanni-t — APPROVED** [review]
Thanks for walking us through the changes. LGTM with the changes from review session

**Review by TamarZanzouri — APPROVED** [review]
Thanks for addressing the changes! looks really good!

**Comment by mcous** [comment]
❌ This merge job failed. This pull request was not merged
Stack job ID: c55nKDJSP4uSRbJYr9kO.
See details on [graphite.dev](https://app.graphite.dev/github/pr/Opentrons/opentrons/11938)

**Pull request merged**